### PR TITLE
feat: state transition conversion

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## What was done?
This PR contains a lot of changes needed for Rs-Drive-Abci to work as the executor.

## How Has This Been Tested?
CI passing.

## Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added or updated relevant unit/integration/functional/e2e tests
- [X] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/rs-dpp/src/data_contract/state_transition/data_contract_update_transition/action.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+use crate::data_contract::state_transition::data_contract_update_transition::DataContractUpdateTransition;
+use crate::data_contract::DataContract;
+use serde::{Deserialize, Serialize};
+
+pub const DATA_CONTRACT_UPDATE_TRANSITION_ACTION_VERSION: u32 = 0;
```
Comment (shumkov) [inline packages/rs-dpp/src/data_contract/state_transition/data_contract_update_transition/action.rs:5]: Why do you need version?

File: packages/rs-dpp/src/data_contract/state_transition/data_contract_create_transition/validation/state/validate_data_contract_create_transition_state.rs [hunk 2]
```diff
@@ -66,17 +67,19 @@ pub async fn validate_data_contract_create_transition_state(
...
-    }
-
-    if maybe_existing_data_contract.is_some() {
-        result.add_error(StateError::DataContractAlreadyPresentError {
-            data_contract_id: state_transition.data_contract.id.to_owned(),
-        })
+    if maybe_existing_data_contract.is_none()
+        || state_transition.get_execution_context().is_dry_run()
+    {
+        let action: DataContractCreateTransitionAction = state_transition.into();
```
Comment (shumkov) [inline packages/rs-dpp/src/data_contract/state_transition/data_contract_create_transition/validation/state/validate_data_contract_create_transition_state.rs:73]: Isn't a mistake?

File: packages/rs-dpp/src/data_contract/state_transition/data_contract_update_transition/validation/state/validate_data_contract_update_transition_state.rs [hunk 3]
```diff
@@ -65,7 +67,8 @@ pub async fn validate_data_contract_update_transition_state(
         .map_err(Into::into)?;
 
     if state_transition.execution_context.is_dry_run() {
```
Comment (shumkov) [inline packages/rs-dpp/src/data_contract/state_transition/data_contract_update_transition/validation/state/validate_data_contract_update_transition_state.rs:69]: All other validation functions now return actions? Doesn't the validation function should only validate?

**Review by shumkov — CHANGES_REQUESTED** [review]

**Review by antouhou — APPROVED** [review]
So we have a very long discussion about whether the function that validates state transition against the state should or shouldn't return a `TransitionAction` as a result, or should it just return the errors and the conversion from state transition into action should be a separate step that's done after the validation. I was a proponent of the idea that it should be two separate steps, and @QuantumExplorer was a proponent of the idea that it should be one step and the conversion of the ST into action should happen inside the validation function. We couldn't agree on that, so we added three more people to the discussion: @shumkov, @fominok, and @shotonoff. At the end of the discussion, we had a vote with three people out of five who voted that the validation function should perform a conversion of the `StateTransition` into an `Action`. So we are going ahead and merging this PR as is. I want also to set up a precedent for the governing of the architectural decisions with this one: although the solution I've proposed has lost the vote, I have no hard feelings about it, and should a disagreement on any other topic arise we should follow the same decision-making protocol, and I expect somebody who lost the consensus vote on a topic to do the same thing as I did :) Thank you, guys!

**Comment by shumkov** [comment]
Yeah, but for the future, I would propose to have ADR for design changes so people don't feel pressure because the code is already written and not having it merged blocks other tasks.

**Pull request merged**